### DISCUSSION TITLE
Support for buffer protocol

### DISCUSSION THREAD

**Discussion by kkrissian-gpfw** [RosettaCommons/binder#215] [Ideas]
Hello,

I have been using Binder to automatically bind several C++ libraries, and I have made some changes to the source code to adapt the tool to specific needs.
One of them was the support for pybind11 buffer_protocol as explained in https://pybind11.readthedocs.io/en/stable/advanced/pycpp/numpy.html with Binder. 

We can consider two ways to use this feature with Binder:
- either bind the whole class, in which case the configuration **_+binder_** is sufficient
- or use an addon on the already bound class with the configuration parameter _**+add_on_binder**_, in which case the parameter py::buffer_protocol() is currently missing in the py::class_ constructor to be able to use .def_buffer() method.

I have made a simple example that uses the buffer_protocol with an addon, and made a few changes in the source code to support a new option **_+buffer_protocol classname_**.

The code is available in our forked repo in branch gp:
* changes in the code are in commit https://github.com/G-P-S/binder/commit/d38ec90f02beedee55d36aca13aba03ef4bfff52
* example and test in commit https://github.com/G-P-S/binder/commit/c9960cfc29fbb735bd8448821d68d8fa15458e89

If you think this feature can be merged into master, I can create a Pull Request for it. 

Thank you and best regards,

Karl Krissian

**Comment by lyskov** [comment]
@kkrissian-gpfw Fantastic! PR adding this feature would certainly be welcome! I briefly looked at the change set and it looks good. One minor point: it would probably be best to move line https://github.com/G-P-S/binder/commit/d38ec90f02beedee55d36aca13aba03ef4bfff52#diff-65d167d0534c5129832f0feba3c3dec474b681d136fbc4f37f7ce5ed523f0217R1185 to config.cpp so it in line with similar config functionality, like say: https://github.com/G-P-S/binder/blob/d38ec90f02beedee55d36aca13aba03ef4bfff52/source/config.cpp#L235 or other `is_xxx` config methods.

Also, if you up to it would probably be best to add simple test case for this functionality. If you would like to do so then please see https://github.com/RosettaCommons/binder/blob/master/test/T31.include_for_class.config for example of how to provide custom config file for test include files.

Thank you and looking forward for PR!

**Reply by kkrissian-gpfw** [reply]
Hi Sergey,

I created the PR https://github.com/RosettaCommons/binder/pull/222

Let me know if it is good, I won't be available next week but if there are changes or comments I will be able to look at them the week after.

Best regards

**Reply by lyskov** [reply]
Fantastic, - PR looks good to me, it is now merged to master. - Thank you for putting it together @kkrissian-gpfw !
